src/exponential_utils.jl

```diff
@@ -0,0 +1,216 @@
+# exponential_utils.jl
+# Contains functions related to the evaluation of scalar/matrix phi functions 
+# that are used by the exponential integrators.
+#
+# TODO: write a version of `expm!` that is non-allocating.
+
+###################################################
+# Dense algorithms
+
+"""
+    phi(z,k[;cache]) -> [phi_0(z),phi_1(z),...,phi_k(z)]
+
+Compute the scalar phi functions for all orders up to k.
+
+The phi functions are defined as
+
+```math
+\\varphi_0(z) = \\exp(z),\\quad \\varphi_k(z+1) = \\frac{\\varphi_k(z) - 1}{z} 
+```
+
+Instead of using the recurrence relation, which is numerically unstable, a 
+formula given by Sidje is used (Sidje, R. B. (1998). Expokit: a software 
+package for computing matrix exponentials. ACM Transactions on Mathematical 
+Software (TOMS), 24(1), 130-156. Theorem 1).
+"""
+function phi(z::T, k::Integer; cache=nothing) where {T <: Number}
+  # Construct the matrix
+  if cache == nothing
+    cache = zeros(T, k+1, k+1)
+  else
+    fill!(cache, zero(T))
+  end
+  cache[1,1] = z
+  for i = 1:k
+    cache[i,i+1] = one(T)
+  end
+  P = Base.LinAlg.expm!(cache)
+  return P[1,:]
+end
+
+"""
+    phimv_dense(A,v,k[;cache]) -> [phi_0(A)v phi_1(A)v ... phi_k(A)v]
+
+Compute the matrix-phi-vector products for small, dense `A`.
+
+The phi functions are defined as
+
+```math
+\\varphi_0(z) = \\exp(z),\\quad \\varphi_k(z+1) = \\frac{\\varphi_k(z) - 1}{z} 
+```
+
+Instead of using the recurrence relation, which is numerically unstable, a 
+formula given by Sidje is used (Sidje, R. B. (1998). Expokit: a software 
+package for computing matrix exponentials. ACM Transactions on Mathematical 
+Software (TOMS), 24(1), 130-156. Theorem 1).
+"""
+function phimv_dense(A, v, k; cache=nothing)
+  w = Matrix{eltype(A)}(length(v), k+1)
+  phimv_dense!(w, A, v, k; cache=cache)
+end
+"""
+    phimv_dense!(w,A,v,k[;cache]) -> w
+
+Non-allocating version of `phimv_dense`.
+"""
+function phimv_dense!(w::AbstractMatrix{T}, A::AbstractMatrix{T}, 
+  v::AbstractVector{T}, k::Integer; cache=nothing) where {T <: Number}
+  @assert size(w, 1) == size(A, 1) == size(A, 2) == length(v) "Dimension mismatch"
+  @assert size(w, 2) == k+1 "Dimension mismatch"
+  m = length(v)
+  # Construct the extended matrix
+  if cache == nothing
+    cache = zeros(T, m+k, m+k)
+  else
+    @assert size(cache) == (m+k, m+k) "Dimension mismatch"
+    fill!(cache, zero(T))
+  end
+  cache[1:m, 1:m] = A
+  cache[1:m, m+1] = v
+  for i = m+1:m+k-1
+    cache[i, i+1] = one(T)
+  end
+  P = Base.LinAlg.expm!(cache)
+  # Extract results
+  @views A_mul_B!(w[:, 1], P[1:m, 1:m], v)
+  for i = 1:k
+    w[:, i+1] = P[1:m, m+i]

```

Comment:
The `P` slice allocates here. `@view` will a little bit as well. Probably easiest to just double loop this one.